src/librbd/internal.cc

```diff
@@ -1943,6 +1943,19 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force) {
 	return r;
       }
 
+      if (!old_format) {
+        r = mirror_image_disable_internal(ictx, true);
+        if (r < 0) {
+          lderr(cct) << "error disabling image mirroring: " << cpp_strerror(r)
+                     << dendl;
+          ictx->owner_lock.put_read();
+          ictx->state->close();
+          return r;
+        }
+
+        // TODO: http://tracker.ceph.com/issues/14421
```

Comment:
I think rbd-mirror will be responsible for propagating the deletion (it should also delete non-primary images if mirroring is disabled).  When the peer rbd-mirror daemon notices that it has a non-primary image owned by a cluster that no longer lists the image in its mirroring directory, it can "force" delete the local, non-primary image.